x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -0,0 +1,176 @@
+    public synchronized Store.MetadataSnapshot openSession(String sessionUUID, IndexShard indexShard) throws IOException {
+        boolean success = false;
+        RestoreContext restore = null;
+        try {
+            if (onGoingRestores.containsKey(sessionUUID)) {
+                logger.debug("not opening new session [{}] as it already exists", sessionUUID);
+                restore = onGoingRestores.get(sessionUUID);
+            } else {
+                logger.debug("opening session [{}] for shard [{}]", sessionUUID, indexShard);
+                if (indexShard.state() == IndexShardState.CLOSED) {
+                    throw new IllegalIndexShardStateException(indexShard.shardId(), IndexShardState.CLOSED,
+                        "cannot open ccr restore session if shard closed");
+                }
+                restore = new RestoreContext(sessionUUID, indexShard, indexShard.acquireSafeIndexCommit());
+                onGoingRestores.put(sessionUUID, restore);
+                openSessionListeners.forEach(c -> c.accept(sessionUUID));
+                HashSet<String> sessions = sessionsForShard.computeIfAbsent(indexShard, (s) ->  new HashSet<>());
+                sessions.add(sessionUUID);
+            }
+            Store.MetadataSnapshot metaData = restore.getMetaData();
+            success = true;
+            return metaData;
+        } finally {
+            if (success ==  false) {
+                onGoingRestores.remove(sessionUUID);
+                IOUtils.closeWhileHandlingException(restore);
+            }
+        }
+    }
+
+    public synchronized void closeSession(String sessionUUID) {
+        closeSessionListeners.forEach(c -> c.accept(sessionUUID));
+        RestoreContext restore = onGoingRestores.remove(sessionUUID);
+        if (restore == null) {
+            logger.info("could not close session [{}] because session not found", sessionUUID);
+            throw new ElasticsearchException("session [" + sessionUUID + "] not found");
+        }
+        IOUtils.closeWhileHandlingException(restore);
+    }
+
+    private class RestoreContext implements Closeable {
+
+        private final String sessionUUID;
+        private final IndexShard indexShard;
+        private final Engine.IndexCommitRef commitRef;
+
+        private RestoreContext(String sessionUUID, IndexShard indexShard, Engine.IndexCommitRef commitRef) {
+            this.sessionUUID = sessionUUID;
+            this.indexShard = indexShard;
+            this.commitRef = commitRef;
+        }
+
+        Store.MetadataSnapshot getMetaData() throws IOException {
+            indexShard.store().incRef();
+            try {
+                return indexShard.store().getMetadata(commitRef.getIndexCommit());
+            } finally {
+                indexShard.store().decRef();
+            }
+        }
+
+        @Override
+        public void close() {
+            removeSessionForShard(sessionUUID, indexShard);
```

Comment:
perhaps assert here that this is always called under lock, i.e.,
```
assert Thread.holdsLock(CcrRestoreSourceService.this);
```

Reply:
Changed